staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+)
+
+// leaseManager is used to manage leases requested from etcd. If a new write
+// needs a lease that has similar expiration time to a previous one, the old
+// lease will be reused to reduce the overhead of etcd, since lease operations
+// are expensive.
+type leaseManager struct {
+	// etcd client used to grant leases
+	client *clientv3.Client
+	// the number of buffered leases, i.e., size of the following mappings
+	size int
+	// the mutex used for the following maps
+	leaseMu sync.Mutex
+	// mapping from lease expiration time in seconds since 1970 to lease id
+	leaseCache map[int64]clientv3.LeaseID
+	// mapping form lease expiration time to last used time in seconds since 1970
+	leaseLastUsedTime map[int64]int64
+	// notification chan to notify that a lease id is granted. It is used
+	// when a lease request find another lease can be used but the lease
+	// id is not yet available.
+	leaseNotification map[int64][]chan clientv3.LeaseID
+	// The period of time in seconds and percent of TTL that each lease is
+	// reused. The minimum of them is used to avoid unreasonably large
+	// numbers. We use var instead of const for testing purposes.
+	leaseReuseDurationSeconds int64
+	leaseReuseDurationPercent float64
+}
+
+// creates a new lease manager using default setting
+func newDefaultLeaseManager(client *clientv3.Client) *leaseManager {
+	return newLeaseManager(client, 8, 60, 0.05)
+}
+
+// creates a new lease manager with the number of buffered leases, lease reuse
+// duration in seconds and percentage. The percentage value x means x*100%.
+func newLeaseManager(client *clientv3.Client, size int, leaseReuseDurationSeconds int64, leaseReuseDurationPercent float64) *leaseManager {
+	return &leaseManager{
+		client: client,
+		size:   size,
+		leaseReuseDurationSeconds: leaseReuseDurationSeconds,
+		leaseReuseDurationPercent: leaseReuseDurationPercent,
+		leaseCache:                make(map[int64]clientv3.LeaseID),
+		leaseLastUsedTime:         make(map[int64]int64),
+		leaseNotification:         make(map[int64][]chan clientv3.LeaseID),
+	}
+}
+
+// SetLeaseReuseDurationSeconds is used for testing purpose. It is used to
+// reduce the extra lease duration to avoid unnecessary timeout in testing.
+func (l *leaseManager) SetLeaseReuseDurationSeconds(duration int64) {
```

Comment:
If this is just for testing purposes - let's not make a public method for it.

Reply:
fixed